### PR TITLE
 [autoscaler] Autoscaler hangs forever on non-zero exit code command

### PR DESCRIPTION
**Pull request by ericl**
## What do these changes do?

Move provider.cleanup() to a finally block to avoid hanging.

## Related issue number

Closes https://github.com/ray-project/ray/issues/3957

### CODE REVIEW

**Comment by ericl** [comment]
cc @ls-daniel @richardliaw would be great to review this for the release

File: python/ray/rllib/setup-rllib-dev.py [hunk 1]
```diff
@@ -11,28 +11,38 @@
 
 import ray
 
-if __name__ == "__main__":
-    rllib_home = os.path.abspath(os.path.join(ray.__file__, "../rllib"))
-    local_home = os.path.abspath(os.path.dirname(__file__))
-    assert os.path.isdir(rllib_home), rllib_home
+
+def do_link(package):
```
Comment (robertnishihara) [inline python/ray/rllib/setup-rllib-dev.py]: the changes in this file are unrelated, right?

Comment (richardliaw) [reply python/ray/rllib/setup-rllib-dev.py]: This is for dev purposes (and are relatively harmless to include in this PR IMO)

**Comment by richardliaw** [comment]
Despite the ugly git diffs, this is essentially only putting `provider.cleanup` in a `finally` section for all ssh cmds and also amending the rllib dev setup script right?

**Comment by ericl** [comment]
Yes

**Review by richardliaw — APPROVED** [review]

**Comment by robertnishihara** [comment]
Failing because of new gym release, see #3973. Should be unrelated.

**Comment by ls-daniel** [comment]
All makes sense to me, apologies for delay.
